pybossa/forms/forms.py

```diff
@@ -684,13 +686,27 @@ class UserPrefMetadataForm(Form):
     review = TextAreaField(
         lazy_gettext('Additional comments'), default="")
 
+    def __init__(self, *args, **kwargs):
+        Form.__init__(self, *args, **kwargs)
+        self.set_can_update(kwargs.get('can_update', (True, None)))
+
     def set_upref_mdata_choices(self):
         upref_mdata_choices = app_settings.upref_mdata.get_upref_mdata_choices()
         self.languages.choices = upref_mdata_choices['languages']
         self.locations.choices = upref_mdata_choices['locations']
         self.timezone.choices = upref_mdata_choices['timezones']
         self.user_type.choices = upref_mdata_choices['user_types']
 
+    def set_can_update(self, can_update_info):
+        self._disabled = self._get_disabled_fields(can_update_info)
+
+    def _get_disabled_fields(self, (can_update, disabled_fields)):
+        if not can_update:
+            return {field: 'Form is not updatable.' for field in self}
+        return {getattr(self, name): reason for name, reason in six.iteritems(disabled_fields or {})}
```

Comment:
`Field` objects have a [`render_kw` property](https://wtforms.readthedocs.io/en/stable/fields.html#wtforms.fields.Field). Is that something we could have used here?

Reply:
I don't see how. We will need to discuss that.

Reply:
Right now the form object maintains the dict with what's going to be on the kwargs. I was just wondering if it is possible to do it at the field level (e.g. setting `render_kw` in the form's `__init__`), so in the templates you only need to change the render_field functions

Reply:
`render_radio_field` does not call the built-in `Field` rendering, so we would have to redesign that function at a minimum to do it this way.

Can we leave this for a future research ticket? The code works well now, and has good separation of concerns between the rendering of disabled widgets on one hand, and the business logic and testing of disabled fields on the other. Doing a complete redesign at this point will prevent me from working on other tickets this sprint, and I want to work on the ANNEX template since it seems more valuable.

Reply:
Build in field rendering is not called, but that property can still be accessed from `field`. In any case, I didn't want you to redesign everything - just checking if form really needs to be concerned with how the field is rendered. It's not important and we can leave this for the future - if we need to add more attributes to the rendered inputs.